design/proposals/ksonnet-lib-simple-constructors.md

```diff
@@ -0,0 +1,109 @@
+
+## Motivation
+
+* ksonnet-lib includes constructors for popular resources. In a few cases, these constructors are incomplete or include items which may not be required. The incompleteness and inconsistencies are a barrier to adoption because developers do not know how to get started in a productive way.
+* ksonnet-lib should be a direct translation of the OpenAPI swagger.json.
+
+## Goal
+
+* Developers should be able to construct objects using the builder pattern rather than constructors.
+* Upstream projects should have a clean slate to build abstraction upon.
+
+## Non Goals
+
+The following are related but not the goals of this specific proposal
+
+* Create upstream projects that implement constructors (e.g. v1 Deployment)
+
+## Proposal
+
+The change would be introducing a new constructor for all resources.
+
+```jsonnet
+group:: {
+  v1:: {
+    local apiVersion = {apiVersion: "group/v1"},
+    kind:: {
+      local kind = {kind: "Kind"},
+      init() :: apiVersion + kind,
+    },
+  },
+}
+```
+
+Many of the existing constructors adhere to this format. This change would make sure all resources have a constructor that only includes the apiVersion and kind.
+
+## User Experience
+
+### Use Cases
+
+To deprecate the existing the existing custom constructors a new constructor will be introduced, `init()`. Currently, to construct a v1beta2 Deployment, you have to do the following:
+
+```jsonnet
+local params = ...;
+local deployment = k.apps.v1beta2.deployment;
+local container = k.apps.v1beta2.deployment.mixin.spec.template.spec.containersType;
+local containerPort = container.portsType;
+local labels = {app: params.name};
+local targetPort = params.containerPort;
+
+local appDeployment = deployment
+  .new(
+    params.name,
+    params.replicas,
+    container
+      .new(params.name, params.image)
+      .withPorts(containerPort.new(targetPort)),
+    labels
+);
+```
+
+This snippet illustrates a few issues:
+
+* The constructor requires items that aren't required by the spec.
+* The constructor requires a specific order.
+
+Instead of the custom constructor, instead you can construct using init using the following:
+
+```jsonnet
+local params = ...;
+local deployment = k.apps.v1beta2.deployment;
+local containersType = k.apps.v1beta2.deployment.mixin.spec.template.spec.containersType;
+local containerPort = container.portsType;
+local labels = {app: params.name};
+local targetPort = params.containerPort;
+
+local port = containerPort
+  .init()
+  .withContainerPort(targetPort)
+
+local container = containersType
+  .init()
+  .withName(params.name)
+  .withImage(params.image)
+  .withPorts(port)
+
+local appDeployment = deployment
+  .init()
+  .mixin.metadata.withName(params.name)
+  .mixin.spec.withReplicas(params.replicas)
+  .mixin.spec.template.spec.withContainers(containers)
+  .mixin.spec.template.metadata.withLabels(labels);
```

Comment:
Maybe this an issue that can be solved with documentation for `ksonnet-lib` (the autogenerated stuff we have planned), but I just wanted to point out that the prefix of `mixin.spec.template.metadata` can be confusing to new users.

If we keep this syntax, we need to point out (e.g. via docs) what a `mixin` is (~ inheritance), and that `spec.template.metadata` are nested fields in the official Deployment spec.

Even if this user knows this, it seems to implicitly require that they are familiar with the K8s spec? Again, perhaps an HTML page with all this info ( + ctrl-F) can be useful, but we should make sure that choice is explicit.

Reply:
This is true. This proposal is trying to make `ksonnet-lib` a ksonnet version of `swagger.json`. All embellishments should have in our yet to be created standard library. It's difficult to use because it allows full access to build the more useful abstractions upstream.

Reply:
Ah ok, that makes sense!